src/org/edx/hadoop/input/ManifestTextInputFormat.java

```diff
@@ -0,0 +1,77 @@
+package org.edx.hadoop.input;
+
+import java.io.*;
+import java.util.List;
+import java.util.ArrayList;
+
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hadoop.mapred.KeyValueTextInputFormat;
+
+
+public class ManifestTextInputFormat extends KeyValueTextInputFormat {
+    public static int RETRY_LIMIT = 5;
+
+    protected FileStatus[] listStatus(JobConf job) throws IOException {
+        FileStatus[] manifests = super.listStatus(job);
+        List<FileStatus> paths = new ArrayList<FileStatus>();
+        for(int i = 0; i < manifests.length; i++) {
+            List<Path> globPaths = this.readManifest(manifests[i].getPath(), job);
+            for (Path globPath : globPaths) {
+                if (globPath != null) {
+                    FileStatus fs = getFileStatus(globPath, job);
+                    if (fs != null)
+                        paths.add(fs);
+                }
+            }
+        }
+        return paths.toArray(new FileStatus[1]);
+    }
+
+    private List<Path> readManifest(Path manifestPath, JobConf job) throws IOException {
+        FileSystem fs = manifestPath.getFileSystem(job);
+        List<Path> paths = new ArrayList<Path>();
+        DataInputStream dataStream = fs.open(manifestPath);
+        try {
+            BufferedReader reader = new BufferedReader(new InputStreamReader(dataStream));
+            String line = reader.readLine();
+            while(line != null) {
+                paths.add(new Path(line));
+                line = reader.readLine();
+            }
+        } finally {
+            dataStream.close();
+        }
+
+        return paths;
+    }
+
+    private FileStatus getFileStatus(Path targetPath, JobConf conf) throws IOException {
+        return getFileStatusWithRetry(targetPath, conf, RETRY_LIMIT);
+    }
+
+    private FileStatus getFileStatusWithRetry(Path targetPath, JobConf conf, int retryCount) throws IOException {
+        if (retryCount <= 0) {
+            LOG.info("Retries exhausted, dropping file: " + targetPath.toUri());
+            throw new IOException("Retries exhausted while querying file: '" + targetPath.toUri() + "'.  Erroring out");
+        }
+
+        FileStatus result = null;
+        try {
+            FileSystem fs = targetPath.getFileSystem(conf);
+            result = fs.getFileStatus(targetPath);
+        } catch (FileNotFoundException e) {
+            LOG.info("File not found: '" + targetPath.toUri() + "'  Ignoring");
+        } catch (IOException e) {
+            LOG.info("Retrying (" + (RETRY_LIMIT - retryCount  + 1) + " of " + RETRY_LIMIT + ") after general exception encountered: " + e.getMessage());
```

Comment:
Should this log statement be put in the else statement below?  When retryCount is 1, this would say "Retrying (5 of 5) after...", but then immediately throw an exception without attempting the retry.  But that raises the question whether we are counting tries or retries.  If the latter, then a retryCount of zero would still try once, but it won't actually.  So retryCount is really something like remainingTryCount.   So maybe we just don't need the "if" below.....